pageserver/src/tenant/ephemeral_file.rs

```diff
@@ -188,51 +181,95 @@ impl EphemeralFile {
         Ok(buf)
     }
 
-    /// Returns the offset at which the first byte of the input was written, for use
-    /// in constructing indices over the written value.
-    ///
-    /// Panics if the write is short because there's no way we can recover from that.
-    /// TODO: make upstack handle this as an error.
     pub(crate) async fn write_raw(
         &self,
         srcbuf: &[u8],
         ctx: &RequestContext,
     ) -> Result<u64, EphemeralFileWriteError> {
-        let (pos, control) = self.write_raw_controlled(srcbuf, ctx).await?;
-        if let Some(control) = control {
-            control.release().await;
-        }
-        Ok(pos)
-    }
+        // serialize writers. this mutex protects the append order and frozen_buffer.
+        // readers do NOT touch this lock so they dont stall on slow disk io.
+        let _write_guard = self.writer_lock.lock().await;
 
-    async fn write_raw_controlled(
```

Comment:
This change effectively re-implements the write path here. That was quite carefully implemented to support pipelining via the flush loop. See `pageserver/src/virtual_file/owned_buffers_io/write/flush.rs`.

I don't see how your implementation supports that.